File: mcp_server/ipc_helpers.py
"""
TestableUIKit MCP Server — Pure helper functions.

HTTP IPC payload construction, response passthrough, and simctl command assembly.
No external dependencies. Testable standalone without Simulator/DemoApp.
"""

import os

DEFAULT_IPC_HOST = "localhost"
DEFAULT_IPC_PORT = 8888

# Env var name constants (Design D: override IPC target for LAN connections)
ENV_IPC_HOST = "TESTABLE_IPC_HOST"
ENV_IPC_PORT = "TESTABLE_IPC_PORT"


def resolve_ipc_host_port(env: dict = None) -> tuple:
    """Resolve the IPC host and port from environment variables.

    Returns the default values (localhost:8888) when no env vars are set.
    Pass a dict as `env` to use it instead of os.environ (for testing).

    Args:
        env: Environment variable dict to use (defaults to os.environ)

    Returns:
        (host: str, port: int) tuple

    Environment variables:
        TESTABLE_IPC_HOST: Target host (default: "localhost")
        TESTABLE_IPC_PORT: Target port as string (default: 8888)
    """
    if env is None:
        env = os.environ
    host = env.get(ENV_IPC_HOST, DEFAULT_IPC_HOST)
    port_str = env.get(ENV_IPC_PORT, str(DEFAULT_IPC_PORT))
    try:
        port = int(port_str)
    except ValueError:
        port = DEFAULT_IPC_PORT
    return host, port


def build_base_url(host: str = DEFAULT_IPC_HOST, port: int = DEFAULT_IPC_PORT) -> str:
    """Build the base URL for the IPC server."""
    return f"http://{host}:{port}"


def build_perform_payload(test_id: str, command_name: str, parameters=None) -> dict:
    """Build the POST /perform request body.

    Args:
        test_id: Component testID (e.g. "scene.auth.loginButton")
        command_name: Command name ("getState" / "tap" / "setProperty", etc.)
        parameters: Command parameters dict (optional)

    Returns:
        {"testID": ..., "commandName": ..., "parameters": ...}
    """
    return {
        "testID": test_id,
        "commandName": command_name,
        "parameters": parameters,
    }


def passthrough_state(response_data: dict) -> dict:
    """Pass through describedState as an opaque dict.

    Does not assume a fixed set of keys (STEP2 forward compatible).
    A thin wrapper that does not modify contents for schema-change resilience.
    """
    return response_data


def build_simctl_screenshot_command(output_path: str) -> list:
    """Assemble the simctl io booted screenshot command.

    Args:
        output_path: Destination file path for the screenshot

    Returns:
        Command list ready to pass to subprocess.run
    """
    return ["xcrun", "simctl", "io", "booted", "screenshot", output_path]


def build_screenshot_url(host: str = DEFAULT_IPC_HOST, port: int = DEFAULT_IPC_PORT) -> str:
    """Build the URL for the GET /screenshot endpoint.

    Args:
        host: Target host (default: "localhost")
        port: Target port (default: 8888)

    Returns:
        URL string in "http://<host>:<port>/screenshot" format
    """
    return f"{build_base_url(host=host, port=port)}/screenshot"


def is_loopback_host(host: str) -> bool:
    """Determine whether the host is a loopback address (pure function).

    Used for the simctl fallback decision:
    Falls back to simctl only when the target is loopback and /screenshot is unreachable.

    Args:
        host: IPC target host string

    Returns:
        True  = loopback ("localhost" / "127.0.0.1" / "::1")
        False = non-loopback such as LAN IP (real device connection, etc.)
    """
    return host in ("localhost", "127.0.0.1", "::1")

File: mcp-swift/Sources/TestableUIKitMCPCore/IPCHelpers.swift
//
//  IPCHelpers.swift
//  TestableUIKitMCPCore
//
//  TestableUIKit MCP Server — pure helper functions (Swift)
//
//  HTTP IPC URL construction, connection resolution, perform payload building,
//  simctl command generation, loopback detection. No external SwiftPM dependencies
//  (Foundation only). Unit-testable without Simulator / DemoApp / HTTP connection.
//
//  Port of mcp_server/ipc_helpers.py (Python). Behavior is matched 1:1.
//

import Foundation

public enum IPCHelpers {

  // ================================================================
  // Constants
  // ================================================================

  /// Default IPC host.
  public static let defaultIPCHost = "localhost"
  /// Default IPC port.
  public static let defaultIPCPort = 8888

  /// Environment variable name for overriding the IPC host (Design D: LAN IPC).
  public static let envIPCHost = "TESTABLE_IPC_HOST"
  /// Environment variable name for overriding the IPC port.
  public static let envIPCPort = "TESTABLE_IPC_PORT"

  // ================================================================
  // resolveIPCHostPort
  // ================================================================

  /// Resolves IPC host and port from environment variables.
  ///
  /// Returns default values (localhost:8888) when the variables are not set.
  /// Falls back to the default port when the port string cannot be parsed as an integer.
  ///
  /// - Parameter env: Environment variable dictionary to read (inject in tests).
  /// - Returns: A (host, port) tuple.
  public static func resolveIPCHostPort(
    env: [String: String]
  ) -> (host: String, port: Int) {
    let host = env[envIPCHost] ?? defaultIPCHost
    let portStr = env[envIPCPort] ?? String(defaultIPCPort)
    let port = Int(portStr) ?? defaultIPCPort
    return (host, port)
  }

  // ================================================================
  // buildBaseURL
  // ================================================================

  /// Builds the IPC server base URL (no trailing slash).
  public static func buildBaseURL(
    host: String = defaultIPCHost,
    port: Int = defaultIPCPort
  ) -> String {
    return "http://\(host):\(port)"
  }

  // ================================================================
  // buildScreenshotURL
  // ================================================================

  /// Builds the URL for the GET /screenshot endpoint.
  public static func buildScreenshotURL(
    host: String = defaultIPCHost,
    port: Int = defaultIPCPort
  ) -> String {
    return "\(buildBaseURL(host: host, port: port))/screenshot"
  }

  // ================================================================
  // buildPerformPayload
  // ================================================================

  /// Builds the POST /perform request body as a JSON string.
  ///
  /// Key order is fixed: testID → commandName → parameters (deterministic).
  /// The caller passes parameters as an already-serialized JSON string.
  /// When nil, the JSON `null` literal is embedded.
  ///
  /// - Parameters:
  ///   - testID: Component testID (e.g. "scene.auth.loginButton").
  ///   - commandName: Command name (e.g. "getState" / "tap" / "setProperty").
  ///   - parametersJSON: Pre-serialized parameters JSON (defaults to null).
  /// - Returns: JSON string of the form `{"testID":...,"commandName":...,"parameters":...}`.
  public static func buildPerformPayload(
    testID: String,
    commandName: String,
    parametersJSON: String? = nil
  ) -> String {
    let params = parametersJSON ?? "null"
    return "{\"testID\":\(jsonEscaped(testID)),"
      + "\"commandName\":\(jsonEscaped(commandName)),"
      + "\"parameters\":\(params)}"
  }

  // ================================================================
  // buildSimctlScreenshotCommand
  // ================================================================

  /// Builds the `simctl io booted screenshot` command arguments (Simulator only).
  ///
  /// - Parameter outputPath: File path where the screenshot will be saved.
  /// - Returns: Argument list suitable for Process / subprocess.
  public static func buildSimctlScreenshotCommand(outputPath: String) -> [String] {
    return ["xcrun", "simctl", "io", "booted", "screenshot", outputPath]
  }

  // ================================================================
  // isLoopbackHost
  // ================================================================

  /// Returns whether the host is a loopback address (pure function).
  ///
  /// Used for simctl fallback decisions: falls back to simctl only when
  /// the host is loopback and the /screenshot endpoint is unreachable.
  ///
  /// - Returns: true = loopback ("localhost" / "127.0.0.1" / "::1").
  public static func isLoopbackHost(_ host: String) -> Bool {
    return host == "localhost" || host == "127.0.0.1" || host == "::1"
  }

  // ================================================================
  // Internal utilities
  // ================================================================

  /// Returns the string as a JSON string literal (double-quoted, escaped).
  static func jsonEscaped(_ s: String) -> String {
    var out = "\""
    for scalar in s.unicodeScalars {
      switch scalar {
      case "\"": out += "\\\""
      case "\\": out += "\\\\"
      case "\n": out += "\\n"
      case "\r": out += "\\r"
      case "\t": out += "\\t"
      default:
        if scalar.value < 0x20 {
          // Control chars → \uXXXX
          let hex = String(scalar.value, radix: 16)
          out += "\\u" + String(repeating: "0", count: 4 - hex.count) + hex
        } else {
          out.unicodeScalars.append(scalar)
        }
      }
    }
    out += "\""
    return out
  }
}
